### ISSUE TITLE
Random errors when bulk-updating document objects via painless script

### ISSUE THREAD

**Issue by consulthys** [elastic/elasticsearch#56414]
**Elasticsearch version** (`bin/elasticsearch --version`): 7.6.1 on Elastic Cloud

**Plugins installed**: N/A

**JVM version** (`java -version`): Elastic Cloud

**OS version** (`uname -a` if on a Unix-like system): Elastic Cloud

**Description of the problem including expected versus actual behavior**:

I'm getting inconsistent and sporadic errors while updating documents (via script) through the `_bulk` API. The issue only happens when trying to update objects (i.e. JSON hash) inside the document.

**Steps to reproduce**:

Here is a way to easily reproduce the problem.

**1. First create a sample index+document**

```
POST test/_doc/1
{
  "hash": {}
}
```

**2. Create a script to be used in the bulk update**

The script simply replaces the existing `hash` field with whatever it receives inside the `params` and adds another field to that hash:

```
POST _scripts/add-data
{
  "script": {
    "lang": "painless",
    "source": """
    ctx._source.hash = params;
    ctx._source.hash.additionalField = 32;
    """
  }
}
```
3. Try to bulk update the document

```
POST test/_doc/_bulk
{"update":{"_id": "1"}}
{"script": {"id":"add-data", "params":{"addedField": 44}}}
```

Now the thing is that sometimes the response is OK

```
{
  "took" : 427,
  "errors" : false,
  "items" : [
    {
      "update" : {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "1",
        "_version" : 2,
        "result" : "updated",
        "_shards" : {
          "total" : 2,
          "successful" : 2,
          "failed" : 0
        },
        "_seq_no" : 1,
        "_primary_term" : 1,
        "status" : 200
      }
    }
  ]
}
```
and sometimes it yields this:
```
{
  "took" : 7,
  "errors" : true,
  "items" : [
    {
      "update" : {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "1",
        "status" : 400,
        "error" : {
          "type" : "illegal_argument_exception",
          "reason" : "failed to execute script",
          "caused_by" : {
            "type" : "script_exception",
            "reason" : "runtime error",
            "script_stack" : [
              "java.base/java.util.Collections$UnmodifiableMap.put(Collections.java:1457)",
              "ctx._source.hash.additionalField = 32;\n    ",
              "                ^---- HERE"
            ],
            "script" : "add-data",
            "lang" : "painless",
            "caused_by" : {
              "type" : "unsupported_operation_exception",
              "reason" : null
            }
          }
        }
      }
    }
  ]
}
```

Now I know that `params` is not modifiable and I'm trying to add a field to it. The error is pretty clear. **But I would expect that it always fails**, not randomly. 

Also worth noting that it makes no difference whether the index has one primary shard or more.

**Comment by consulthys** [comment]
@dimitris-athanasiou I'm not sure how this fixes the issue I've reported above... Are you sure about the issue number ?

**Comment by consulthys** [comment]
@dimitris-athanasiou I'm not sure how this fixes the issue I've reported above... Are you sure about the issue number ? Wasn't it #56419 ?

**Comment by hendrikmuhs** [comment]
@consulthys I think you are right, this bug has been accidentally closed via a commit message

**Comment by stu-elastic** [comment]
This should always fail, we'll try to reproduce.

**Comment by consulthys** [comment]
@stu-elastic Note that I can still reproduce this in 8.13.3

**Comment by rjernst** [comment]
params is unfortunately not read-only. In _some_ cases it can become read-only (eg passing script object over the wire), but making it read-only across all script contexts is a technically breaking change.

**Comment by consulthys** [comment]
I tested again on 8.17.3 and it now fails consistently.

**Comment by rjernst** [comment]
@consulthys would you consider this fixed then? My reading of the original issue is that the bug was the inconsistency, not that there was an error.

**Comment by consulthys** [comment]
@rjernst yes, that's correct, I originally didn't mind the error, just the inconsistency of it being sometimes reported sometimes not. Feel free to close it.
